### PR TITLE
Add OpenPOWER platform detection

### PR DESCRIPTION
**Pull request by runlevel5**
## Context

As suggested by Mike, I create this PR to add OpenPOWER CPU detection into `Hardware::CPU` class.

## Changes

* Add optimisation flag for PowerPC64
* Enhance the condition check for tigerbrew

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb [hunk 1]
```diff
@@ -4,16 +4,22 @@ module Hardware
...
         armv8:   "-march=armv8-a",
-      }.freeze
+        ppc64le: "-mcpu=powerpc64le",
+      }
 
       def optimization_flags
-        OPTIMIZATION_FLAGS_LINUX
+        OPTIMIZATION_FLAGS_LINUX.tap do |flags|
+          if ppc?
+            flags[:native] = "-mcpu=#{Homebrew::EnvConfig.arch}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/hardware/cpu.rb:19]: Why does PPC use `mcpu` here?

File: Library/Homebrew/hardware.rb [hunk 2]
```diff
@@ -7,28 +7,35 @@ class CPU
...
         core:    "-march=prescott",
         armv6:   "-march=armv6",
         armv8:   "-march=armv8-a",
+        ppc64:   "-mcpu=powerpc64",
+        ppc64le: "-mcpu=powerpc64le",
       }.freeze
 
       def optimization_flags
-        OPTIMIZATION_FLAGS
+        OPTIMIZATION_FLAGS_LINUX.tap do |flags|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:24]: Shouldn't reference `OPTIMIZATION_FLAGS_LINUX` in this non-Linux file

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb (hunk 1 above)
Comment (runlevel5) [reply Library/Homebrew/extend/os/linux/hardware/cpu.rb]: `-march` is the x86 options. For ppc, we use `-mcpu`

Ref: https://gcc.gnu.org/onlinedocs/gcc/RS_002f6000-and-PowerPC-Options.html#RS_002f6000-and-PowerPC-Options

File: Library/Homebrew/hardware.rb [hunk 3]
```diff
@@ -7,28 +7,35 @@ class CPU
...
         armv8:   "-march=armv8-a",
+        ppc64:   "-mcpu=powerpc64",
+        ppc64le: "-mcpu=powerpc64le",
       }.freeze
 
       def optimization_flags
-        OPTIMIZATION_FLAGS
+        OPTIMIZATION_FLAGS.tap do |flags|
+          flags[:native] = if ppc?
+            "-mcpu=native"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:26]: Still don't understand the difference for PPC here and why this can't be part of the hash above instead?

Comment (runlevel5) [reply Library/Homebrew/hardware.rb:26]: If I understand your question right, are you asking why this line could not be incorporated to the [line 14](https://github.com/Homebrew/brew/pull/7465/files/63ec2409c39ba12662f744b56c5fa85b665feed9#diff-da65ce29d8f934c7bbb4631f3e1fe672L14)? (that is `native: "-march=native -mcpu=native"`?

Comment (MikeMcQuaid) [reply Library/Homebrew/hardware.rb:26]: Ok, I see now. Can this instead be split into a function in `Homebrew/hardware` that takes a parameter e.g. `native` or `Homebrew::EnvConfig.arch` and returns `-mcpu=#{param}` or `-march=#{param}`

File: Library/Homebrew/hardware.rb [hunk 4]
```diff
@@ -7,28 +7,31 @@ class CPU
     INTEL_32BIT_ARCHS = [:i386].freeze
     INTEL_64BIT_ARCHS = [:x86_64].freeze
     PPC_32BIT_ARCHS   = [:ppc, :ppc32, :ppc7400, :ppc7450, :ppc970].freeze
-    PPC_64BIT_ARCHS   = [:ppc64].freeze
+    PPC_64BIT_ARCHS   = [:ppc64, :ppc64le, :ppc970].freeze
 
     class << self
       OPTIMIZATION_FLAGS = {
-        native:  "-march=native",
         nehalem: "-march=nehalem",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:14]:
```suggestion
        native:  arch_flag("native"),
        nehalem: "-march=nehalem",
```

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb [hunk 5]
```diff
@@ -4,16 +4,19 @@ module Hardware
   class CPU
     class << self
       OPTIMIZATION_FLAGS_LINUX = {
-        native:  "-march=#{Homebrew::EnvConfig.arch}",
         nehalem: "-march=nehalem",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/hardware/cpu.rb:7]:
```suggestion
        native:  arch_flag(Homebrew::EnvConfig.arch),
        nehalem: "-march=nehalem",
```

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb [hunk 6]
```diff
@@ -4,16 +4,19 @@ module Hardware
...
         armv8:   "-march=armv8-a",
+        ppc64:   "-mcpu=powerpc64",
+        ppc64le: "-mcpu=powerpc64le",
       }.freeze
 
       def optimization_flags
-        OPTIMIZATION_FLAGS_LINUX
+        OPTIMIZATION_FLAGS_LINUX.dup.tap do |flags|
+          flags[:native] = arch_flag(Homebrew::EnvConfig.arch)
+        end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/hardware/cpu.rb:19]:
```suggestion
        OPTIMIZATION_FLAGS_LINUX
```

File: Library/Homebrew/hardware.rb [hunk 7]
```diff
@@ -7,28 +7,31 @@ class CPU
...
         armv8:   "-march=armv8-a",
+        ppc64:   "-mcpu=powerpc64",
+        ppc64le: "-mcpu=powerpc64le",
       }.freeze
 
       def optimization_flags
-        OPTIMIZATION_FLAGS
+        OPTIMIZATION_FLAGS.dup.tap do |flags|
+          flags[:native] = arch_flag("native")
+        end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:26]:
```suggestion
        OPTIMIZATION_FLAGS
```

File: Library/Homebrew/hardware.rb [hunk 8]
```diff
@@ -110,17 +115,41 @@ def ppc?
...
         []
       end
 
       def feature?(name)
         features.include?(name)
       end
+
+      def arch_flag(arch)
+        ppc? ? "-mcpu=#{arch}" : "-march=#{arch}"
+      end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:152]:
```suggestion
      def arch_flag(arch)
        return "-mcpu=#{arch}" if ppc?
        
        "-march=#{arch}"
      end
```

File: Library/Homebrew/install.rb [hunk 9]
```diff
@@ -10,13 +10,18 @@ module Install
...
-      case Hardware::CPU.type
-      when :ppc
+      if OS.mac? && Hardware::CPU.ppc?
         abort <<~EOS
           Sorry, Homebrew does not support your computer's CPU architecture.
-          For PPC support, see:
+          For PowerPC Macs (PPC32/PPC64BE) support, see:
             #{Formatter.url("https://github.com/mistydemeo/tigerbrew")}
         EOS
+      elsif OS.linux? && Hardware::CPU.ppc64le?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/install.rb:19]: This should live in a new `Library/Homebrew/extend/os/linux/install.rb`. See the other files in that directory for the pattern.

File: Library/Homebrew/install.rb [hunk 10]
```diff
@@ -10,13 +10,18 @@ module Install
     module_function
 
     def check_cpu
-      case Hardware::CPU.type
-      when :ppc
+      if OS.mac? && Hardware::CPU.ppc?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/install.rb:13]:
```suggestion
      if Hardware::CPU.ppc?
```

File: Library/Homebrew/install.rb [hunk 11]
```diff
@@ -10,13 +10,18 @@ module Install
...
         abort <<~EOS
           Sorry, Homebrew does not support your computer's CPU architecture.
-          For PPC support, see:
+          For PowerPC Macs (PPC32/PPC64BE) support, see:
```
Comment (MikeMcQuaid) [inline Library/Homebrew/install.rb:16]:
```suggestion
          For PowerPC Mac (PPC32/PPC64BE) support, see:
```

File: Library/Homebrew/install.rb (hunk 10 above)
Comment (runlevel5) [reply Library/Homebrew/install.rb:13]: I get it, so this is by default for mac. What do you think about moving this to `Library/Homebrew/extend/os/mac/install.rb`?

Comment (MikeMcQuaid) [reply Library/Homebrew/install.rb:13]: We tend to keep the Mac defaults in `install.rb` and then override for Linux.

File: Library/Homebrew/hardware.rb (hunk 4 above)
Comment (runlevel5) [reply Library/Homebrew/hardware.rb:14]: This won't work because this constant is defined at singleton class level.

Comment (MikeMcQuaid) [reply Library/Homebrew/hardware.rb:15]: Yeh, of course. How about: get rid of the `OPTIMIZATION_FLAGS` constant (it's unused) and move it all into the `optimization_flags` flag method instead.

Comment (runlevel5) [reply Library/Homebrew/hardware.rb:15]: If we want to call the class method `arch_flag`, this constant would need to be moved to class level

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb [hunk 12]
```diff
@@ -3,17 +3,17 @@
...
-        nehalem: "-march=nehalem",
-        core2:   "-march=core2",
-        core:    "-march=prescott",
-        armv6:   "-march=armv6",
-        armv8:   "-march=armv8-a",
-      }.freeze
-
       def optimization_flags
-        OPTIMIZATION_FLAGS_LINUX
+        {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/hardware/cpu.rb:7]: What about reusing the non-Linux values and only adding to the existing hash where needed?

File: Library/Homebrew/hardware.rb [hunk 13]
```diff
@@ -7,28 +7,28 @@ class CPU
...
+        {
+          native:  arch_flag("native"),
+          nehalem: "-march=nehalem",
+          core2:   "-march=core2",
+          core:    "-march=prescott",
+          armv6:   "-march=armv6",
+          armv8:   "-march=armv8-a",
+          ppc64:   "-mcpu=powerpc64",
+          ppc64le: "-mcpu=powerpc64le",
+        }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:23]:
```suggestion
        }.freeze
```

File: Library/Homebrew/hardware.rb [hunk 14]
```diff
@@ -7,28 +7,28 @@ class CPU
...
       end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:24]:
```suggestion
      end
      alias generic_optimization_flags optimization_flags
```

then you could reuse it in the Linux code above and modify the value

File: Library/Homebrew/hardware.rb [hunk 15]
```diff
@@ -7,28 +7,28 @@ class CPU
...
-        nehalem: "-march=nehalem",
-        core2:   "-march=core2",
-        core:    "-march=prescott",
-        armv6:   "-march=armv6",
-        armv8:   "-march=armv8-a",
-      }.freeze
-
       def optimization_flags
-        OPTIMIZATION_FLAGS
+        {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/hardware.rb:14]:
```suggestion
        @optimization_flags ||= {
```

maybe that's overkill, though?

**Comment by runlevel5** [comment]
@MikeMcQuaid the PR is now ready for review

File: Library/Homebrew/hardware.rb (hunk 15 above)
Comment (runlevel5) [reply Library/Homebrew/hardware.rb:14]: Abit but I think it's good practice to memoize

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb [hunk 16]
```diff
@@ -3,17 +3,17 @@
...
+          native:  arch_flag(Homebrew::EnvConfig.arch),
+          nehalem: "-march=nehalem",
+          core2:   "-march=core2",
+          core:    "-march=prescott",
+          armv6:   "-march=armv6",
+          armv8:   "-march=armv8-a",
+          ppc64:   "-mcpu=powerpc64",
+          ppc64le: "-mcpu=powerpc64le",
+        }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/hardware/cpu.rb:16]:
```suggestion
        @optimization_flags ||= begin
          flags = generic_optimization_flags.dup
          flags[:native] = arch_flag(Homebrew::EnvConfig.arch)
          flags
        end
```

maybe something like this? Haven't checked that it works or is correct!

File: Library/Homebrew/extend/os/linux/install.rb [hunk 17]
```diff
@@ -14,6 +14,16 @@ module Install
...
     ].freeze
 
+    def check_cpu
+      return unless Hardware::CPU.ppc64le?
+
+      abort <<~EOS
+        Sorry, Homebrew does not support your computer's CPU architecture.
+        For OpenPOWER Linux (PPC64LE) support, see:
+          #{Formatter.url("https://github.com/homebrew-ppc64le/brew")}
+      EOS
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/install.rb:24]:
```suggestion
      return if (Hardware::CPU.intel? && Hardware::CPU.is_64_bit? || Hardware::CPU.arm?)

      message = "Sorry, Homebrew does not support your computer's CPU architecture!"
      if Hardware::CPU.ppc64le?
        message += <<~EOS
          For OpenPOWER Linux (PPC64LE) support, see:
            #{Formatter.url("https://github.com/homebrew-ppc64le/brew")}
        EOS
      end
      abort message
```

File: Library/Homebrew/extend/os/linux/install.rb [hunk 18]
```diff
@@ -14,6 +14,19 @@ module Install
       "/system/bin/linker",
     ].freeze
 
+    def check_cpu
+      return if (Hardware::CPU.intel? && Hardware::CPU.is_64_bit? || Hardware::CPU.arm?)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/install.rb:18]:
```suggestion
      return if (Hardware::CPU.intel? && Hardware::CPU.is_64_bit?) || Hardware::CPU.arm?
```

My bad on style here sorry!

**Comment by MikeMcQuaid** [comment]
@runlevel5 I'm happy with all the code now and it looks like it will pass all tests! Can you double-check that all the code as-is will work as desired/expected on your fork? Thanks!

**Comment by MikeMcQuaid** [comment]
Thanks so much for your contribution (hopefully one of many)! Without people like you submitting PRs we couldn't run this project. You rock, @runlevel5!

**Pull request merged**